File: src/main/java/io/Cbuild/gnu_make_functions/compiler/overwrite/compile_wordlist.java
package io.Cbuild.gnu_make_functions.compiler.overwrite;

import io.Cbuild.cBuildIR;
import io.Cbuild.cbuildException;
import io.Cbuild.cbuildParser;
import io.Cbuild.gnu_make_functions.compiler.compile_fn;
import io.Cbuild.gnu_make_functions.compiler.util;
import io.Cbuild.gnu_make_functions.type;

public class compile_wordlist extends compile_fn {

    @Override
    public cBuildIR.FunctionIR compile(cbuildParser.FunctionContext ctx, type.MakeFunction function) {
        cBuildIR.FunctionIR ir = super.compile(ctx, function);
        util.cleanWS(ir);

        if (ir.args.getFirst().parts.size() != 1) {
            throw new cbuildException(
                    cbuildException.ErrorType.SEMANTIC,
                    "wordlist: expected exactly one expression as the first argument",
                    ctx.start.getLine(),
                    ctx.start.getCharPositionInLine() + 1
            );
        }

        if (ir.args.get(1).parts.size() != 1) {
            throw new cbuildException(
                    cbuildException.ErrorType.SEMANTIC,
                    "wordlist: expected exactly one expression as the second argument",
                    ctx.start.getLine(),
                    ctx.start.getCharPositionInLine() + 1
            );
        }

        return ir;
    }

}

File: src/gnu-make-functions/compiler/overwrite/compile_wordlist.ts
import { FunctionIR } from "@compiler/ir.js";
import { FunctionContext } from "@parser/cbuildParser.js";
import {
  CbuildException,
  ErrorType,
  MachineCode,
} from "@src/cbuild-exception.js";
import type { MakeFunction } from "@gnu-make-functions/type.js";
import { compile_fn } from "@gnu-make-functions/compiler/compile_fn.js";
import { util } from "@gnu-make-functions/compiler/util.js";

export class compile_wordlist extends compile_fn {
  public override compile(
    ctx: FunctionContext,
    func: MakeFunction,
  ): FunctionIR {
    const ir = super.compile(ctx, func);
    util.cleanWS(ir);

    if (ir.args[0].parts.length !== 1) {
      throw CbuildException.from({
        errorType: ErrorType.SEMANTIC,
        message:
          "buildFile: wordlist: expected exactly one expression as the first argument",
        row: ctx.start?.line || 0,
        column: (ctx.start?.column || 0) + 1,
        machineCode: MachineCode.FUNCTION_COMPILATION_ERROR,
      });
    }

    if (ir.args[1].parts.length !== 1) {
      throw CbuildException.from({
        errorType: ErrorType.SEMANTIC,
        message:
          "buildFile: wordlist: expected exactly one expression as the second argument",
        row: ctx.start?.line || 0,
        column: (ctx.start?.column || 0) + 1,
        machineCode: MachineCode.FUNCTION_COMPILATION_ERROR,
      });
    }

    return ir;
  }
}
